xds/src/main/java/io/grpc/xds/OrcaOobUtil.java

```diff
@@ -0,0 +1,627 @@
+      }
+
+      void stopRpc(String msg) {
+        if (orcaRpc != null) {
+          orcaRpc.cancel(msg);
+          orcaRpc = null;
+        }
+        if (retryTimer != null) {
+          retryTimer.cancel();
+          retryTimer = null;
+        }
+      }
+
+      boolean isRetryTimerPending() {
+        return retryTimer != null && retryTimer.isPending();
+      }
+
+      @Override
+      public String toString() {
+        return MoreObjects.toStringHelper(this)
+            .add("disabled", disabled)
+            .add("orcaRpc", orcaRpc)
+            .add("reportingConfig", overallConfig)
+            .add("connectivityState", state)
+            .toString();
+      }
+
+      private class OrcaReportingStream extends ClientCall.Listener<OrcaLoadReport> {
+
+        private final ClientCall<OrcaLoadReportRequest, OrcaLoadReport> call;
+        private final Stopwatch stopwatch;
+        private boolean callHasResponded;
+
+        OrcaReportingStream(Channel channel, Stopwatch stopwatch) {
+          call =
+              checkNotNull(channel, "channel")
+                  .newCall(OpenRcaServiceGrpc.getStreamCoreMetricsMethod(), CallOptions.DEFAULT);
+          this.stopwatch = checkNotNull(stopwatch, "stopwatch");
+        }
+
+        void start() {
+          stopwatch.reset().start();
+          call.start(this, new Metadata());
+          call.sendMessage(
+              OrcaLoadReportRequest.newBuilder()
+                  .setReportInterval(Durations.fromNanos(overallConfig.getReportIntervalNanos()))
+                  .build());
+          call.halfClose();
+          call.request(1);
+        }
+
+        @Override
+        public void onMessage(final OrcaLoadReport response) {
+          syncContext.execute(
+              new Runnable() {
+                @Override
+                public void run() {
+                  if (orcaRpc == OrcaReportingStream.this) {
+                    handleResponse(response);
+                  }
+                }
+              });
+        }
+
+        @Override
+        public void onClose(final Status status, Metadata trailers) {
+          syncContext.execute(
+              new Runnable() {
+                @Override
+                public void run() {
+                  if (orcaRpc == OrcaReportingStream.this) {
+                    orcaRpc = null;
+                    handleStreamClosed(status);
+                  }
+                }
+              });
+        }
+
+        void handleResponse(OrcaLoadReport response) {
+          callHasResponded = true;
+          backoffPolicy = null;
+          subchannelLogger.log(ChannelLogLevel.DEBUG, "Received an ORCA report: {0}", response);
+          for (OrcaOobReportListener listener : listeners) {
+            listener.onLoadReport(response);
+          }
+          call.request(1);
+        }
+
+        void handleStreamClosed(Status status) {
+          if (Objects.equal(status.getCode(), Code.UNIMPLEMENTED)) {
```

Comment:
I'm curious why this is not `if (status.getCode() == Code.UNIMPLEMENTED)`. And also about `Objects.equal(state.getState(), READY)`. They are enum types.

Reply:
I would dare to say, they are equally good here. :flushed::flushed:
So I will leave it as is.